resources/media-controls.js

```diff
@@ -0,0 +1,381 @@
+            return positionTextNode.textContent;
+          },
+          set: v => {
+            positionTextNode.textContent = positionFormat.replace("#1", v);
+          }
+        },
+        duration: {
+          get: () => {
+            return durationSpan.textContent;
+          },
+          set: v => {
+            durationSpan.textContent = v ? durationFormat.replace("#2", v) : "";
+          }
+        },
+        show: {
+          value: (currentTime, duration) => {
+            const self = this.elements.positionDurationBox;
+            if (self.position != currentTime) {
+              self.position = currentTime;
+            }
+            if (self.duration != duration) {
+              self.duration = duration;
+            }
+            self.classList.remove("hidden");
+          }
+        }
+      });
+
+      // Add event listeners.
+      this.mediaEvents = [
+        "play",
+        "pause",
+        "ended",
+        "volumechange",
+        "loadeddata",
+        "loadstart",
+        "timeupdate",
+        "progress",
+        "playing",
+        "waiting",
+        "canplay",
+        "canplaythrough",
+        "seeking",
+        "seeked",
+        "emptied",
+        "loadedmetadata",
+        "error",
+        "suspend"
+      ];
+      this.mediaEvents.forEach(event => {
+        this.media.addEventListener(event, this);
+      });
+
+      this.controlEvents = [
+        { el: this.elements.playPauseButton, type: "click" },
+        { el: this.elements.volumeSwitch, type: "click" },
+        { el: this.elements.volumeLevel, type: "input" }
+      ];
+      this.controlEvents.forEach(({ el, type }) => {
+        el.addEventListener(type, this);
+      });
+
+      // Create state transitions.
+      //
+      // It exposes one method per transition. i.e. this.pause(), this.play(), etc.
+      // For each transition, we check that the transition is possible and call
+      // the `onStateChange` handler.
+      for (let name in TRANSITIONS) {
+        if (!TRANSITIONS.hasOwnProperty(name)) {
+          continue;
+        }
+        this[name] = () => {
+          const from = this.state;
+
+          // Checks if the transition is valid in the current state.
+          if (!TRANSITIONS[name][from]) {
+            const error = `Transition "${name}" invalid for the current state "${from}"`;
+            console.error(error);
+            throw new Error(error);
+          }
+
+          const to = TRANSITIONS[name][from];
+
+          if (from == to) {
+            return;
+          }
+
+          // Transition to the next state.
+          this.state = to;
+          this.onStateChange(from);
+        };
+      }
+
+      // Set initial state.
+      this.state = this.media.paused ? PAUSED : PLAYING;
+      this.onStateChange(null);
+    }
+
+    cleanup() {
+      this.mutationObserver.disconnect();
+      this.mediaEvents.forEach(event => {
+        this.media.removeEventListener(event, this);
```

Comment:
Don't we need something Ike `this[name]` as the listener parameter here?

Reply:
AFAICT this is the way to remove the listeners here. Note that we are using the `handleEvent` method for all events, and the corresponding [addEventListeners](https://github.com/servo/servo/pull/23208/files#diff-454017e7187d4e177929eecd1c4d4d3cR177) for this case.

Reply:
Oh, I missed that. Great.